src/System.CommandLine.Subsystems/Subsystems/Annotations/AnnotationStorageExtensions.cs

```diff
@@ -0,0 +1,142 @@
+    // symbols are no longer reachable. Although this is unlikely to happen in a CLI app, it is important not to create
+    // unexpected, unfixable, unbounded memory leaks in apps that construct multiple grammars/pipelines.
+    //
+    // The main use case for System.CommandLine is for a CLI app to construct a single annotated grammar in its entry point,
+    // construct a pipeline using that grammar, and use the pipeline/grammar only once to parse its arguments. However, it
+    // is important to have well defined and reasonable threading behavior so that System.CommandLine does not behave in
+    // surprising ways when used in more advanced cases:
+    //
+    // * There may be multiple threads constructing and using completely independent grammars/pipelines. This happens in
+    //   our own unit tests, but might happen e.g. in a multithreaded data processing app or web service that uses
+    //   System.CommandLine to process inputs.
+    //
+    // * The grammar/pipeline are reentrant; they do not store they do not store internal state, and may be used to parse
+    //   input multiple times. As this is the case, it is reasonable to expect a grammar/pipeline instance to be
+    //   constructed in one thread then used in multiple threads. This might be done by the aforementioned web service or
+    //   data processing app.
+    //
+    // The thread-safe behavior of ConditionalWeakTable ensures this works as expected without us having to worry about
+    // taking locks directly, even though the instance is on a static field and shared between all threads. Note that
+    // thread local storage is not useful for this, as that would create unexpected behaviors where a grammar constructed
+    // in one thread would be missing its annotations when used in another thread.
+    //
+    // However, while getting values from ConditionalWeakTable is lock free, setting values internally uses an expensive
+    // lock, so it is not ideal to store all individual annotations directly in the ConditionalWeakTable. This is especially
+    // true as we do not want the common case of the CLI app entrypoint to have its performance impacted by multithreading
+    // support more than absolutely necessary.
+    //
+    // Instead, we have a single static ConditionalWeakTable that maps each CliSymbol to an AnnotationStorage dictionary,
+    // which is lazily created and added to the ConditionalWeakTable a single time for each CliSymbol. The individual
+    // annotations are stored in the AnnotationStorage dictionary, which uses no locks, so is fast, but is not safe to be
+    // modified from multiple threads.
+    //
+    // This is fine, as we will have the following well-defined threading behavior: an annotated grammar and pipeline may
+    // only be constructed/modified from a single thread. Once the grammar/pipeline instance is fully constructed, it may
+    // be safely used from multiple threads.
```

Comment:
We might want to add that once constructed, it is not safe to modify the grammar if it is in use.